helix-core/src/main/java/org/apache/helix/task/WorkflowDispatcher.java

```diff
@@ -165,7 +165,9 @@ public void updateWorkflowStatus(String workflow, WorkflowConfig workflowCfg,
     RuntimeJobDag runtimeJobDag = _clusterDataCache.getTaskDataCache().getRuntimeJobDag(workflow);
     if (runtimeJobDag != null) {
       for (String inflightJob : runtimeJobDag.getInflightJobList()) {
-        processJob(inflightJob, currentStateOutput, bestPossibleOutput, workflowCtx);
+        if (isJobReadyToExecute(workflowCtx, inflightJob)) {
```

Comment:
I am OK to fix it right now. But please put a note or TODO here. Adding non-scheduled job to inflight job list make people conceptually confused.

Reply:
@dasahcc I discussed with @alirezazamani offline. I feel that we just neglected all the checking required before calling `processJob` on inflight jobs. "Inflight" in this sense doesn't mean that a job is necessarily running; rather, it means that the job has been passed over onto the Controller and now out of RuntimeJobDAG's responsibility. Otherwise we end up doing a lot of "passing jobs back and forth" for every failure/delay situation that will complicate the code/logic even further.

So in short, it was a simple bug that we didn't check before doing processJob here.